### ISSUE TITLE
misc.h, memcpy_s assert() fails

### ISSUE THREAD

**Issue by GeorgeMatveev** [weidai11/cryptopp#125]
Compiles fine but crashes with:

crypt: /usr/include/cryptopp/misc.h:304: void CryptoPP::memcpy_s(void*, size_t, const void*, size_t): Assertion `dest != __null' failed.
Aborted (core dumped)

Debugging leads to:

(gdb) up
#5  0x00007ffff79898ff in CryptoPP::AlgorithmParametersBase::GetVoidValue(char const*, std::type_info const&, void*) const () from /usr/lib/libcrypto++.so.9
(gdb)

Could be a good idea to make Base64Encoder.IsolatedInitialize(params) simpler to ease debugging!

**Comment by noloader** [comment]
@GeorgeMatveev - what is the platform, what compiler, and what's the origin of the source code (distro, 5.6.3 ZIP, GitHub, etc).

TLDR; - can you recompile the library and see if the issue is still present?

-----

We recently cleared cast-alignment warnings (see [Issue 122](http://github.com/weidai11/cryptopp/issues/122)). A couple of the changes had the potential to need a recompile. I *thought* I avoided the recompile requirement, but I may have introduced one.

We also had two small issues in the ***`SecBlock`*** class, but they were cleared. Neither issue was present in the self tests. However, when we used [Gcove](https://gcc.gnu.org/onlinedocs/gcc/Gcov.html) to add missing test cases, we found the issues. I seem to recall one of the issues showed up when using ***`memcpy_s`***.

**Comment by GeorgeMatveev** [comment]
Been using cryptopp code from GigHub compiled using:

geo@geo-Lenovo-G580:~/kelisec/aes256$ g++ --version
g++ (Ubuntu 4.8.4-2ubuntu1~14.04) 4.8.4

code in the app:

Base64Encoder encoder;
AlgorithmParameters params = MakeParameters(Name::EncodingLookupArray(),(const byte *)ALPHABET);
encoder.IsolatedInitialize(params);

Compiles fine but crashes.
Using several ctors/functions in MakeParameters() does not make debugging easier
and needs to be split in 2 more lines imho.
For now using CryptoPP::AES::Encryption which works fine and gives expected results.

Hope it will help to improve very useful library.

Cheers,
George Matveev
www.matveev.se

**Comment by noloader** [comment]
Looking at frame 5 from the original post:

>     #5 0x00007ffff79898ff in CryptoPP::AlgorithmParametersBase::GetVoidValue(char const, std::type_info const&,
>     void) const () from /usr/lib/libcrypto++.so.9

It looks like you are compiling against your cloned repo, but runtime linking against the distro's shared object.

This is easiest enough to fix.

    g++ -DNDEBUG -g2 -O2 -I . foo.cpp ./libcryptopp.a -o foo.exe

The ***`-I .`*** says to find header in PWD. The ***`./libcryptopp.a`*** is the static library, and it avoids the problem you are encountering.
